### ISSUE TITLE
[REVIEW]: funsies: A minimalist, distributed and dynamic workflow engine

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3274]
**Submitting author:** @clavigne (<a href="http://orcid.org/0000-0003-2778-1866">Cyrille Lavigne</a>)
**Repository:** <a href="https://github.com/aspuru-guzik-group/funsies" target ="_blank">https://github.com/aspuru-guzik-group/funsies</a>
**Version:** 0.8.1
**Editor:** @diehlpk
**Reviewer:** @gflofst, @vijaysm
**Archive:** <a href="https://doi.org/10.5281/zenodo.5528107" target="_blank">10.5281/zenodo.5528107</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/e5c1e6bf39474021f548f0f0aefd9cb8/status.svg)](https://joss.theoj.org/papers/e5c1e6bf39474021f548f0f0aefd9cb8)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/e5c1e6bf39474021f548f0f0aefd9cb8"><img src="https://joss.theoj.org/papers/e5c1e6bf39474021f548f0f0aefd9cb8/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/e5c1e6bf39474021f548f0f0aefd9cb8/status.svg)](https://joss.theoj.org/papers/e5c1e6bf39474021f548f0f0aefd9cb8)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/aspuru-guzik-group/funsies" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@camillescott & @gflofst & @vijaysm, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @diehlpk know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/aspuru-guzik-group/funsies">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@clavigne) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/aspuru-guzik-group/funsies">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@clavigne) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @camillescott, @gflofst it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.29 s (270.4 files/s, 54489.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
HTML                            10            639              0           6867
Python                          57           1304           1324           4879
Markdown                         6             75              0            380
TeX                              1             27              0            294
YAML                             1              0              0             40
INI                              2              3              0             29
Bourne Shell                     1             11             23             14
Mako                             1              0              0              8
-------------------------------------------------------------------------------
SUM:                            79           2059           1347          12511
-------------------------------------------------------------------------------


Statistical information for the repository 'faecd364a3f68a5fe5aff86a' was
gathered on 2021/05/11.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Cyrille Lavigne                317         23835          14394          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Cyrille Lavigne            7316           30.7          2.2               12.40
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.26434/chemrxiv.13008500.v1 may be a valid DOI for title: Automatic discovery of chemical reactions using imposed activation
- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by diehlpk** [comment]
> ```
> Reference check summary (note 'MISSING' DOIs are suggestions that need verification):
> 
> OK DOIs
> 
> - 10.1063/5.0006002 is OK
> - 10.1063/5.0004835 is OK
> - 10.1063/5.0007045 is OK
> - 10.1063/5.0004997 is OK
> - 10.1063/5.0005188 is OK
> - 10.1063/1.5144261 is OK
> - 10.1016/j.matt.2021.02.017 is OK
> - 10.1145/1330017.1330019 is OK
> - 10.1093/bioinformatics/bts480 is OK
> - 10.12688/f1000research.29032.2 is OK
> - 10.1145/2807591.2807623 is OK
> - 10.1007/10968987_3 is OK
> - 10.1007/978-3-319-63688-7_19 is OK
> 
> MISSING DOIs
> 
> - 10.26434/chemrxiv.13008500.v1 may be a valid DOI for title: Automatic discovery of chemical reactions using imposed activation
> - 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters
> 
> INVALID DOIs
> 
> - None
> ```

@clavigne please add the missing DOIs to the paper and recompile the paper.

**Comment by clavigne** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: :point_left:

**Comment by clavigne** [comment]
I added the DOI for the Chemrxiv pre-print. The other paper, (which is available here https://research.google/pubs/pub62/ ) does not have a DOI, as far as I can tell. 

The ACM which compiles the USENIX proceedings doesn't have the OSDI '04 one https://dl.acm.org/conference/osdi/proceedings .

**Comment by diehlpk** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by diehlpk** [comment]
@clavigne Can you please add this one missing DOI

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

**Comment by clavigne** [comment]
> 
> 
> I added the DOI for the Chemrxiv pre-print. The other paper, (which is available here https://research.google/pubs/pub62/ ) does not have a DOI, as far as I can tell.
> 
> The ACM which compiles the USENIX proceedings doesn't have the OSDI '04 one https://dl.acm.org/conference/osdi/proceedings .

@diehlpk as above, there are no DOI for this reference, as far as I know.

**Comment by diehlpk** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by diehlpk** [comment]
@clavigne 

> 
> MISSING DOIs
> 
> - 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

@whedon still complains about this missing DOI. Can you have a look?

**Comment by whedon** [comment]
:wave: @gflofst, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @camillescott, please update us on how your review is going (this is an automated reminder).

**Comment by gflofst** [comment]
added issue for making community guidelines more apparent. Continuing other testing.

**Comment by gflofst** [comment]
Everything works for me. The last bit of testing on Windows getting an up to date Redis working is going to take time. As it is the last bit, and everything else is good, I am ok signing off on it for now and I'll raise an issue later if it isn't working. Windows is a marginal platform for this field, but it is what I have for my desktop machine for testing.

**Comment by diehlpk** [comment]
@camillescott How is your review going?

**Comment by diehlpk** [comment]
@camillescott How is your review going?

**Comment by diehlpk** [comment]
@whedon remind @camillescott in one week

**Comment by whedon** [comment]
Reminder set for @camillescott in one week

**Comment by whedon** [comment]
:wave: @camillescott, please update us on how your review is going (this is an automated reminder).

**Comment by vijaysm** [comment]
The installation for funsies was smooth on both OSX and Linux. But I couldn't get pip to install the redis-server on OSX. Had to resort to homebrew solutions. Is this expected ? Using `Python==3.8.10`.

I was able to run the example workflows just fine without issues after that initial installation hiccup.

**Comment by vijaysm** [comment]
@clavigne  The mergesort example is interesting. However, running it gives me several error messages like below. It would be better to catch these exceptions so that the example output is cleaner.
```
 op:bf485b | runner raised!
Traceback (most recent call last):

  File "/usr/local/lib/python3.8/dist-packages/funsies/_run.py", line 186, in run_op
    out_data: out_data_t = runner(funsie, input_data)  # type:ignore
                           │      │       └ {'in0': b'[6]'}
                           │      └ Funsie(how=<FunsieHow.subdag: 2>, what='SAC|split|funsies_mergesort.<locals>.<lambda>|funsies_mergesort.<locals>.<lambda>', i...
                           └ <function run_subdag_funsie at 0x7f8dc5690280>

  File "/usr/local/lib/python3.8/dist-packages/funsies/_subdag.py", line 81, in run_subdag_funsie
    outfun = fun(decoded)
             │   └ {'in0': [6]}
             └ <function sac.<locals>.__sac at 0x7f8dc54b3670>

  File "/usr/local/lib/python3.8/dist-packages/funsies/dynamic.py", line 64, in __sac
    split_data = [constant_artefact(db, d) for d in split_fun(*args)]
                  │                 │               │          └ [[6]]
                  │                 │               └ <function split at 0x7f8dc54b38b0>
                  │                 └ Redis<ConnectionPool<Connection<host=localhost,port=6379,db=0>>>
                  └ <function constant_artefact at 0x7f8dc56d9dc0>

  File "mergesort.py", line 18, in split
    raise RuntimeError()

RuntimeError
 op:bf485b | DONE: runner raised exception.
```
If the error messages are not expected, then it should be fixed. If this is the expected behavior (?), then perhaps an explanation in documentation on why there is a `RuntimeError` being raised, and how it gets handled internally would be helpful. I can replicate this on both OSX and Linux.

**Comment by diehlpk** [comment]
@clavigne can you please look at the reviewer's comments?

**Comment by clavigne** [comment]
Hi @vijaysm , thank you for the careful review.

> The installation for funsies was smooth on both OSX and Linux. But I couldn't get pip to install the redis-server on OSX. Had to resort to homebrew solutions. Is this expected ? Using `Python==3.8.10`.
> 
> I was able to run the example workflows just fine without issues after that initial installation hiccup.

I do not have access to an OSX machine to test it, so I'm not sure if this is expected. The Redis installation through pip is from this package https://github.com/pywharf/redis-server which states support for `macosx_10_15_x86_64`. At this time, Redis 4.x or above is supported by funsies. I've added this information to the Readme, as well as the description of how to install it through Homebrew.

**Comment by clavigne** [comment]
@vijaysm as to your second comment,

> 
> 
> @clavigne The mergesort example is interesting. However, running it gives me several error messages like below. It would be better to catch these exceptions so that the example output is cleaner.
> 
> ```
>  op:bf485b | runner raised!
> Traceback (most recent call last):
> 
>   File "/usr/local/lib/python3.8/dist-packages/funsies/_run.py", line 186, in run_op
>     out_data: out_data_t = runner(funsie, input_data)  # type:ignore
>                            │      │       └ {'in0': b'[6]'}
>                            │      └ Funsie(how=<FunsieHow.subdag: 2>, what='SAC|split|funsies_mergesort.<locals>.<lambda>|funsies_mergesort.<locals>.<lambda>', i...
>                            └ <function run_subdag_funsie at 0x7f8dc5690280>
> 
>   File "/usr/local/lib/python3.8/dist-packages/funsies/_subdag.py", line 81, in run_subdag_funsie
>     outfun = fun(decoded)
>              │   └ {'in0': [6]}
>              └ <function sac.<locals>.__sac at 0x7f8dc54b3670>
> 
>   File "/usr/local/lib/python3.8/dist-packages/funsies/dynamic.py", line 64, in __sac
>     split_data = [constant_artefact(db, d) for d in split_fun(*args)]
>                   │                 │               │          └ [[6]]
>                   │                 │               └ <function split at 0x7f8dc54b38b0>
>                   │                 └ Redis<ConnectionPool<Connection<host=localhost,port=6379,db=0>>>
>                   └ <function constant_artefact at 0x7f8dc56d9dc0>
> 
>   File "mergesort.py", line 18, in split
>     raise RuntimeError()
> 
> RuntimeError
>  op:bf485b | DONE: runner raised exception.
> ```
> 
> If the error messages are not expected, then it should be fixed. If this is the expected behavior (?), then perhaps an explanation in documentation on why there is a `RuntimeError` being raised, and how it gets handled internally would be helpful. I can replicate this on both OSX and Linux.

The raised errors are used for control flow. They terminate the mergesort recursion dynamically without having to know in advance the required recursion depth. I'm updating the example to make this more clear, and to make it raise a more self-explanatory exception.

(By the way, the exception is in fact caught, it's just displayed by the worker in its log file for analysis. This is *usually* the desired behaviour: the worker doesn't die but does report that there is a problem. However, as you point out, there isn't really any problem here, so the raising a `RuntimeError()` is rather poor design!)

**Comment by vijaysm** [comment]
@clavigne Thanks for quickly updating README. 

> I do not have access to an OSX machine to test it, so I'm not sure if this is expected. The Redis installation through pip is from this package https://github.com/pywharf/redis-server which states support for macosx_10_15_x86_64. At this time, Redis 4.x or above is supported by funsies. I've added this information to the Readme, as well as the description of how to install it through Homebrew.

I didn't drill down on the reasons why pip didn't install on OSX, but showing alternative workflows to get the TPLs installed is helpful for new users. The installation information looks complete enough now.

**Comment by vijaysm** [comment]
> The raised errors are used for control flow. They terminate the mergesort recursion dynamically without having to know in advance the required recursion depth. I'm updating the example to make this more clear, and to make it raise a more self-explanatory exception.

Got it. I assumed the runtime error was intentional, but the log messages with trace information would confuse users, and lead them to worry that either the example is broken, or their build/installation is incompatible. So displaying this in a better, more user-friendly way would be better. 

You should also update README.md or appropriate documentation for funsies to describe how runtime errors are caught cleanly and redirected to the log file instead of workers quitting prematurely. This is an important feature for all real workflows and should be highlighted explicitly IMO.

Also, is there a separate documentation/example page for funsies ? I don't see one. There is a link to source docs and link to the recipes folder, but something like a user guide will add more value here as well. Bulk of what would go in a user document is already there in README.md, but separating this perhaps into its own space (github.io ?) will allow you to add more details as needed in terms of describing usage scenarios, details on examples, envisioned future extensions etc.

**Comment by vijaysm** [comment]
Keyboard shortcut accidentally closed the issue :-) I re-opened it again.

**Comment by vijaysm** [comment]
@clavigne Can you add other workflow engines like Snakemake that are routinely used for distributed systems ? I am aware of [Fabsim](https://fabsim3.readthedocs.io/en/latest/index.html) which is a flexible workflow engine, though targeted at multiscale apps. It may be a bit more complex in terms of usage though. Another one is [Galaxy](https://training.galaxyproject.org/training-material/topics/admin/tutorials/connect-to-compute-cluster/tutorial.html). Your reference Molder et al., (2021) is for Snakemake, and I do not explicitly see any citations to "other" engines. It would be helpful to provide more context on what else is available with a similar feature set, and why funsies is different in the statement of need.

**Comment by danielskatz** [comment]
There's also [Parsl](http://parsl-project.org) but I'm writing this more for your information than asking you to mention it, since I work on it.

**Comment by diehlpk** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# All commands can be run on a non-default branch, to do this pass a custom 
# branch name by following the command with `from branch custom-branch-name`.
# For example:

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a dry run of accepting the paper and depositing with Crossref
@whedon recommend-accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by diehlpk** [comment]
@whedon remind @vijaysm  in 2 weeks

**Comment by whedon** [comment]
Reminder set for @vijaysm  in 2 weeks

**Comment by clavigne** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: :point_left:

**Comment by clavigne** [comment]
Hi @vijaysm thanks again for the thorough review and I apologize for the delay in my reply. Hopefully I can provide satisfactory responses to your raised points, but please advise if you feel that this is not the case.

> Got it. I assumed the runtime error was intentional, but the log messages with trace information would confuse users, and lead them to worry that either the example is broken, or their build/installation is incompatible. So displaying this in a better, more user-friendly way would be better.

I changed the example to make the reporting (somewhat) more user-friendly. It now raises `StopRecursion` instead of `RuntimeError` and the behaviour is mentioned specifically in https://github.com/aspuru-guzik-group/funsies/blob/master/recipes/silly/README.md .

> You should also update README.md or appropriate documentation for funsies to describe how runtime errors are caught cleanly and redirected to the log file instead of workers quitting prematurely. This is an important feature for all real workflows and should be highlighted explicitly IMO.

That's an excellent point. I added this information to https://github.com/aspuru-guzik-group/funsies/blob/master/README.md . A more complete treatment will have to wait until the documentation is online (as described below) and I've finished designing the error handling. In particular, I'm still not entirely sure whether the current error handling is sufficiently ergonomic; it's the main part I want to stabilize at this time.

> Also, is there a separate documentation/example page for funsies ? I don't see one. There is a link to source docs and link to the recipes folder, but something like a user guide will add more value here as well. Bulk of what would go in a user document is already there in README.md, but separating this perhaps into its own space (github.io ?) will allow you to add more details as needed in terms of describing usage scenarios, details on examples, envisioned future extensions etc.

I agree, a separate documentation page is sorely needed. I haven't had the time to dive into it yet, and have been waiting on comments from other users about some of the ergonomics. I'll write a more complete documentation when I can free some time for it, hopefully before Fall.

> @clavigne Can you add other workflow engines like Snakemake that are routinely used for distributed systems ? I am aware of Fabsim which is a flexible workflow engine, though targeted at multiscale apps. It may be a bit more complex in terms of usage though. Another one is Galaxy. Your reference Molder et al., (2021) is for Snakemake, and I do not explicitly see any citations to "other" engines. It would be helpful to provide more context on what else is available with a similar feature set, and why funsies is different in the statement of need.

Yes! I cited Molder et al. because it contains the most thorough review and classification of those systems that I'm aware of. I've modified the paper to make this clearer, 

> Scientific workflow management systems, many of which are compiled in [@awesome_pipeline] and systematically reviewed in [@molder_sustainable_2021], are easier to set up on high-performance computing clusters, but are tuned to the needs of specific disciplines, such as bioinformatics or machine learning. 

Please advise if this solution is satisfactory. I didn't want to make the paper too long by reviewing what is (at this point) a very large number of alternatives (it's a bit of a crowded space).

Thanks again!

**Comment by clavigne** [comment]
@danielskatz There's also Parsl but I'm writing this more for your information than asking you to mention it, since I work on it.

Thanks! I actually wasn't aware of Parsl when I embarked on this project, and by the time I did become aware of it, I already had done a fair bit of work. Parsl is quite similar to funsies (python-based, similar architecture, free licenses), albeit significantly more mature. There are probably interesting opportunities for collaboration and code sharing between the two projects (or at least for code I can borrow into funsies!)

**Comment by danielskatz** [comment]
If you (very optionally) want to mention it, feel free to do so - we like citations, particularly of https://doi.org/10.1145/3307681.3325400 !

And if you want to discuss collaboration, we're happy to do that too - best might be to join the Parsl slack via http://parsl-project.org/support.html

**Comment by diehlpk** [comment]
Hey @vijaysm how is your review going?

**Comment by vijaysm** [comment]
@diehlpk @clavigne I apologize about the delay. My primary machine got fried due to some coffee spillage recently, and so my productivity has been shot. I should be able to read the paper containing the new changes soon, and will finish the full review by next week.

**Comment by diehlpk** [comment]
@vijaysm Sorry to hear that your primary machine got fried.

**Comment by vijaysm** [comment]
@clavigne The article reads well and provides a clear description of the funsies package. Just one minor comment: consider adding a DOI for the citation Dean, J., Ghemawat, S. (2004). I could not find one for that version when I checked quickly. But the 2008 version of the paper has one. https://doi.org/10.1145/1327452.1327492

I should be able to wrap up the rest of my review in the next couple of days.

**Comment by diehlpk** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by diehlpk** [comment]
@clavigne Can you please add the missing DOI to the paper?

**Comment by clavigne** [comment]
@diehlpk As I mentioned previously ( https://github.com/openjournals/joss-reviews/issues/3274#issuecomment-848041524 ) this specific article doesn't have DOI. The reviewer @vijaysm recommends instead citing the later review on MapReduce ( https://doi.org/10.1145/1327452.1327492 ) as opposed to the primary article ( https://www.usenix.org/legacy/events/osdi04/tech/dean.html ). This is an acceptable solution if citations must absolutely be associated with DOIs, but I'd rather cite the primary literature if possible which is why I did not make the change previously.

I'm willing to change the reference of course, but I feel like there is a bug in the JOSS software/process if references not accompanied with DOIs are rejected. It's not uncommon to cite DOI-less resources such as government reports or older papers (such as this one).

**Comment by diehlpk** [comment]
> @diehlpk As I mentioned previously ( https://github.com/openjournals/joss-reviews/issues/3274#issuecomment-848041524 ) this specific article doesn't have DOI. The reviewer @vijaysm recommends instead citing the later review on MapReduce ( https://doi.org/10.1145/1327452.1327492 ) as opposed to the primary article ( https://www.usenix.org/legacy/events/osdi04/tech/dean.html ). This is an acceptable solution if citations must absolutely be associated with DOIs, but I'd rather cite the primary literature if possible which is why I did not make the change previously.
> 
> I'm willing to change the reference of course, but I feel like there is a bug in the JOSS software/process if references not accompanied with DOIs are rejected. It's not uncommon to cite DOI-less resources such as government reports or older papers (such as this one).

@clavigne Ok, I misread the comment. Everything is fine.

**Comment by diehlpk** [comment]
@vijaysm How is your review going?

**Comment by diehlpk** [comment]
@vijaysm How is your review going?

**Comment by vijaysm** [comment]
Another small note is that I was not able to run your instructions above on Cori. For example, after installing funsies through `pip`, `start-funsies` was not callable. I thought perhaps this was installed with pip, but it's not available in my `$PATH`.

However, I was able to test the installation and runs on Cori at NERSC. The example workflow provided [here](https://github.com/aspuru-guzik-group/funsies/tree/3b4425457d51f9a90d3163f696bf8d8b51779d5d/recipes/slurm-conformers) worked cleanly with appropriate changes to the `slurm-submit.sh` file. @clavigne It would be good to add a link to the Slurm submit example in README.md or even provide a separate subsection there to talk about using funsies for workflows on LCF machines or clusters.

@diehlpk I've completed my review and I am satisfied with the software documentation and the accompanying paper. I was also able to test funsies installation and examples on a variety of different platforms without any issues. 

I also started writing a new example to orchestrate one of my workflows in Climate data analysis, but did not make enough progress on that the last two weeks. That was another reason for the delay - but I think I'll continue that experiment separately and raise issues (if any) directly in the repository in the next couple of weeks.

**Comment by diehlpk** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: :point_left:

**Comment by diehlpk** [comment]
@clavigne Could you please check the blue names with the question mark at the end?

It seems that they are rendered as links, however, these are not links. Are these GitHub usernames? If so, the @ could be the problem. 

```
We acknowledge testing by early users Cher-Tian Ser ((chertianser?)), Kjell Jorner96
((kjelljorner?)) and Gabriel dos Passos Gomes ((gabegomes?)). CL also thanks Chris97
Crebolder ((ccrebolder?)) f
```

**Comment by diehlpk** [comment]
@clavigne 

```
We intend to maintain funsies and of course welcome collaborations from contributors93
around the world.
```

I think it could be nice to refer to 

https://github.com/aspuru-guzik-group/funsies/blob/master/CONTRIBUTING.md

in the sentence above.

**Comment by diehlpk** [comment]
@clavigne I just got power back and will read your paper the next days, and we should be able to finalize your paper by the end of this week.

**Comment by diehlpk** [comment]
@clavigne I can approve the paper from the editorial perspective, please work on the two tasks above. After that I will have some final look.

**Comment by diehlpk** [comment]
@clavigne I would like to finish this paper soon, since I want to take the next paper.

**Comment by clavigne** [comment]
Hi @diehlpk sorry for the delay, I was on holiday and then I moved to a new positions. I'm implementing those changes right now.

**Comment by clavigne** [comment]
@whedon generate

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by clavigne** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: :point_left:

**Comment by clavigne** [comment]
@diehlpk done, thank you for your patience.

**Comment by diehlpk** [comment]
@clavigne Thanks for the edits, please create

- [ ] A new version including all the requested changes by the reviewers
- [ ] Upload this version to Zenodo or FigShare. Note that the authors and the title have to match with the paper

and post the version and DOI here, so I can add them to the paper and recommend it for acceptance.

**Comment by diehlpk** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by clavigne** [comment]
Hi @diehlpk , version 0.8.1 is the final version of the software for publication.

The Zenodo DOI is 10.5281/zenodo.5528107 (url: https://doi.org/10.5281/zenodo.5528107 )

Thank you,
Cyrille

**Comment by diehlpk** [comment]
@whedon set 0.8.1 as version

**Comment by whedon** [comment]
OK. 0.8.1 is the version.

**Comment by diehlpk** [comment]
@whedon set 10.5281/zenodo.5528107 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5528107" target="_blank">10.5281/zenodo.5528107</a> is the archive.

**Comment by diehlpk** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by diehlpk** [comment]
@whedon generate pdf

**Comment by diehlpk** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2611

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2611, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by diehlpk** [comment]
@gflofst, @vijaysm thanks again for your review.

**Comment by danielskatz** [comment]
@clavigne - I've suggested some small changes in https://github.com/aspuru-guzik-group/funsies/pull/5
(Note that removing the abstracts simply cleans up the XML file that we submit to Crossref - it has no effect on generated paper).

Let me know when you've merged these changes, and we can proceed to the final publishing step.

Also, and just FYI, this work is interesting to read in the context of the [Parsl project](http://parsl-project.org) to which I contribute.

**Comment by danielskatz** [comment]
👋 @clavigne - note that this is basically ready to proceed, once you merge the PR...

**Comment by clavigne** [comment]
@danielskatz @diehlpk Thanks a lot for your help! I merged in the changes. I think the paper is ready for publication.

**Comment by clavigne** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03274/joss.03274/10.21105.joss.03274.pdf) :page_facing_up: :point_left:

**Comment by clavigne** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
I'm sorry @clavigne, I'm afraid I can't do that. That's something only editor-in-chiefs are allowed to do.

**Comment by danielskatz** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.26434/chemrxiv.13008500.v2 is OK
- 10.1063/5.0006002 is OK
- 10.1063/5.0004835 is OK
- 10.1063/5.0007045 is OK
- 10.1063/5.0004997 is OK
- 10.1063/5.0005188 is OK
- 10.1063/1.5144261 is OK
- 10.1016/j.matt.2021.02.017 is OK
- 10.1145/1330017.1330019 is OK
- 10.1093/bioinformatics/bts480 is OK
- 10.12688/f1000research.29032.2 is OK
- 10.1145/2807591.2807623 is OK
- 10.1007/10968987_3 is OK
- 10.1007/978-3-319-63688-7_19 is OK

MISSING DOIs

- 10.21276/ijre.2018.5.5.4 may be a valid DOI for title: MapReduce: Simplified Data Processing on Large Clusters

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2640

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2640, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@clavigne - sorry, there's one more thing I missed initially and then just caught: https://github.com/aspuru-guzik-group/funsies/pull/6

**Comment by clavigne** [comment]
@danielskatz no need to apologize for making the paper better :D Thanks again!

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1444367233237585926) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2641
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03274](https://doi.org/10.21105/joss.03274)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
Note that at least, for me, the DOI is not yet resolving, so I'll keep this open until it does, the close the issue.

**Comment by danielskatz** [comment]
Congratulations to @clavigne (Cyrille Lavigne) and co-author!!

And thanks to @gflofst and @vijaysm for reviewing, and @diehlpk for editing!
We couldn't do this without you!

**Comment by arfon** [comment]
DOI is resolving now.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03274/status.svg)](https://doi.org/10.21105/joss.03274)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03274">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03274/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03274/status.svg
   :target: https://doi.org/10.21105/joss.03274
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03274/status.svg)](https://doi.org/10.21105/joss.03274)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by danielskatz** [comment]
The DOI now resolves, so we're done!!

**Comment by xyluo25** [comment]
Hi @clavigne, I hope you’re doing well and enjoying a productive week. I recently submitted the pyufunc paper to JOSS (just yesterday), and I was delighted to see that JOSS recommended you as an ideal reviewer given your expertise. I know how busy schedules can be, so I would be incredibly grateful if you could kindly consider reviewing the manuscript. Your insights and feedback would mean a great deal to me and help ensure pyufunc meets the highest standards.

Please let me know if you have any questions or need additional information. Thank you so much for your time and consideration. I truly appreciate it.

**Comment by gflofst** [comment]
I am happy to review. I'll have a bit more time in July (3 weeks in a row at home!).

**Comment by xyluo25** [comment]
> I am happy to review. I'll have a bit more time in July (3 weeks in a row at home!).

Thank you @gflofst! Wish you have a wonderful day!

**Comment by kthyng** [comment]
@xyluo25 Please leave the reviewer recruiting to your editor.

**Comment by clavigne** [comment]
@xyluo25 I'd be happy to do so but 

@kthyng should I wait for an editor message? I haven't gotten any. My academic email has lapsed now since I moved to industry.

**Comment by kthyng** [comment]
@clavigne You are welcome to volunteer to review on the pre-review page if you are so inclined!

**Comment by gflofst** [comment]
Is this for a new release of the software? I remember reviewing this now that I dig in.

**Comment by kthyng** [comment]
@gflofst The review in the present issue is completed and you were a review on it, but no further action is needed here. Thanks!
